test/blackbox-tests/test-cases/watching/dir-target-promotion.t

```diff
@@ -0,0 +1,170 @@
+Test directory target promotion in file-watching mode.
+
+  $ . ./helpers.sh
+
+  $ echo '(lang dune 3.0)' > dune-project
+  $ mkdir test; cd test
+  $ cat > dune-project <<EOF
+  > (lang dune 3.0)
+  > (using directory-targets 0.1)
+  > EOF
+  $ cat > dune <<EOF
+  > (rule
+  >   (mode promote)
+  >   (deps src (sandbox always))
+  >   (targets (dir d1))
+  >   (action (bash "mkdir -p d1/d2; cp src d1/a; cp src d1/b; cp src d1/d2/c")))
+  > EOF
+
+  $ echo -n "*" > src
+  $ start_dune
+
+  $ build d1
+  Success
+  $ cat d1/a d1/b d1/d2/c
+  ***
+
+Change [src] and rebuild.
+
+  $ echo -n "+" > src
+  $ build d1
+  Success
+  $ cat d1/a d1/b d1/d2/c
+  +++
+
+Remove a file and rebuild.
+
+  $ rm d1/a
+  $ build d1
+  Success
+  $ cat d1/a d1/b d1/d2/c
+  +++
+
+Remove a directory and rebuild.
+
+  $ rm -rf d2
+  $ build d1
+  Success
+  $ cat d1/a d1/b d1/d2/c
+  +++
+
+Modify a file and rebuild.
+
+  $ echo -n "*" > d1/b
+  $ cat d1/a d1/b d1/d2/c
+  +*+
+  $ build d1
+  Success
+  $ cat d1/a d1/b d1/d2/c
+  +++
+
+Replace a directory with a file and rebuild.
+
+  $ rm -rf d1/d2
+  $ touch d1/d2
+  $ build d1
+  Failure
+
+Delete the problematic file and retry.
+
+  $ rm d1/d2
+  $ build d1
+  Success
+  $ cat d1/a d1/b d1/d2/c
+  +++
+
+Add some unexpected files and directories and check that Dune deletes them.
+
+# CR-someday amokhov: This scenario works but Dune currently produces some
+# spurious errors (see below) that go away after retriggering. We should fix this.
+
+  $ mkdir -p d1/unexpected-dir-1
+  $ mkdir -p d1/d2/unexpected-dir-2
+  $ touch d1/unexpected-file-1
+  $ touch d1/unexpected-dir-1/unexpected-file-2
+  $ touch d1/d2/unexpected-file-3
+  $ build d1
+  Success
+  $ cat d1/a d1/b d1/d2/c
+  +++
+  $ ls d1 | grep unexpected
+  [1]
+  $ ls d1/d2 | grep unexpected
+  [1]
+
+We're done.
+
+  $ stop_dune
+  Success, waiting for filesystem changes...
+  Success, waiting for filesystem changes...
+  Success, waiting for filesystem changes...
+  Success, waiting for filesystem changes...
+  Success, waiting for filesystem changes...
+  File "dune", line 1, characters 0-151:
+  1 | (rule
+  2 |   (mode promote)
+  3 |   (deps src (sandbox always))
+  4 |   (targets (dir d1))
+  5 |   (action (bash "mkdir -p d1/d2; cp src d1/a; cp src d1/b; cp src d1/d2/c")))
+  Error: Cannot promote files to "d1/d2".
```

Comment:
As discussed off-line, Dune should probably just delete `d1/d2` here. This is inline with what we do for promoted files where we erase any user modification.

We should also test the opposite: having a directory instead of a file.

Reply:
Agreed.